bitcoin/src/blockdata/fee_rate.rs

```diff
@@ -4,102 +4,158 @@ use core::fmt;
         // 1 vb == 4 wu
         // 1 sat/vb == 1/4 sat/wu
         // sat_vb sat/vb * 1000 / 4 == sat/kwu
-        Some(FeeRate(sat_vb.checked_mul(1000 / 4)?))
+
+        // NB remove indirection after impl const support for ops
+        if let Some(per_kwu) = per_vb.checked_mul(1000 / 4) {
+            Some(FeeRate { per_kwu })
+        } else {
+            None
+        }
     }
 
-    /// Constructs `FeeRate` from satoshis per virtual bytes without overflow check.
-    pub const fn from_sat_per_vb_unchecked(sat_vb: u64) -> Self { FeeRate(sat_vb * (1000 / 4)) }
+    /// Constructs `FeeRate` from an amount per virtual bytes without overflow check.
+    pub const fn from_per_vb_unchecked(per_vb: Amount) -> Self {
+        FeeRate { per_kwu: Amount::from_sat(per_vb.to_sat() * (1000 / 4)) }
+    }
 
-    /// Returns raw fee rate.
+    /// Returns the fee rate per kilo weight unit.
     ///
     /// Can be used instead of `into()` to avoid inference issues.
-    pub const fn to_sat_per_kwu(self) -> u64 { self.0 }
+    pub const fn to_per_kwu(self) -> Amount { self.per_kwu }
 
-    /// Converts to sat/vB rounding down.
-    pub const fn to_sat_per_vb_floor(self) -> u64 { self.0 / (1000 / 4) }
+    /// Converts to amount per virtual byte rounding down.
+    pub const fn to_per_vb_floor(self) -> Amount {
+        // NB remove indirection after impl const support for ops
+        Amount::from_sat(self.per_kwu.to_sat() / (1000 / 4))
+    }
 
-    /// Converts to sat/vB rounding up.
-    pub const fn to_sat_per_vb_ceil(self) -> u64 { (self.0 + (1000 / 4 - 1)) / (1000 / 4) }
+    /// Converts to amount per virtual byte rounding up.
+    pub const fn to_per_vb_ceil(self) -> Amount {
+        // NB remove indirection after impl const support for ops
+        Amount::from_sat((self.per_kwu.to_sat() + 1000 / 4 - 1) / (1000 / 4))
+    }
 
     /// Checked multiplication.
     ///
     /// Computes `self * rhs` returning `None` if overflow occurred.
-    pub fn checked_mul(self, rhs: u64) -> Option<Self> { self.0.checked_mul(rhs).map(Self) }
+    pub const fn checked_mul(self, rhs: u64) -> Option<Self> {
+        // NB remove indirection after impl const support for ops
+        match self.per_kwu.checked_mul(rhs) {
+            Some(v) => Some(Self::from_per_kwu(v)),
+            None => None,
+        }
+    }
 
     /// Checked division.
     ///
     /// Computes `self / rhs` returning `None` if `rhs == 0`.
-    pub fn checked_div(self, rhs: u64) -> Option<Self> { self.0.checked_div(rhs).map(Self) }
+    pub const fn checked_div(self, rhs: u64) -> Option<Self> {
+        // NB remove indirection after impl const support for ops
+        match self.per_kwu.checked_div(rhs) {
+            Some(v) => Some(Self::from_per_kwu(v)),
+            None => None,
+        }
+    }
+
+    /// Constructs `FeeRate` from satoshis per 1000 weight units.
+    #[deprecated(since = "0.31.0", note = "use the denomination-independent methods instead")]
```

Comment:
Hmm, I didn't think this through too much, but this might be able to land in 0.30.1 already instead of 0.31.0. I never understood how people used this tag because you can hardly know beforehand at what version this lands...

Reply:
I wonder if we should do `since = "100.0.0"` or something and then have our "release check" CI job require that there aren't any `since` values that exceed the actual release version, so we remember to fix them before release.

Anyway not related to this PR, we don't need to deal with it here :)

Reply:
That would be fancy. And would need some syntax tree inspection right? I guess having a convention to always use `100.0.0` and then grep for `100.0.0` would also work a lot easier :)

Reply:
>And would need some syntax tree inspection right?

Heh, I guess ... though I suspect we could do it with `grep` and it'd be "good enough". Then once I get some `cargo doc`-based tooling this sort of thing would be super easy.

Reply:
I notice the field is freeform, so I can put there `NEXT_RELEASE`. That'd be easy enough to grep for in the release CI, I guess?

Reply:
Gave this a shot here https://github.com/rust-bitcoin/rust-bitcoin/pull/1791. If that would get merge, I can rebase this MR and use `NEXT_RELEASE`.